### PR TITLE
Integrate Woodard et al. (2021) permafrost code and parameters

### PR DESCRIPTION
**Pull request by bpbond**
This PR integrates the permafrost code from [Woodard et al. 2021](https://doi.org/10.5194/gmd-14-4751-2021) by @dawnlwoodard based on original code by @ashiklom . It has been entirely re-implemented for Hector v3, so while it follows the same logic there are some differences:

* We use Boost for lognormal numerics instead of R - see #512 
* `static_c` isn't tracked separately; we simply use the `f_static` value to implicitly calculate it in respiration calculations

## Performance for SSP 2.45

permafrost

Corresponding figure (not exactly, but many of same outputs; note this is using an older version of the model and the RCPs) from Woodard et al.:

Screenshot 2022-12-22 at 10 29 58 AM

Permafrost loss (a couple hundred Pg C) and temperature change (~0.4 K) look pretty good.

Thaw curve:

f_frozen

Screenshot 2022-12-22 at 10 32 12 AM

## Other

* Permafrost is off (i.e. `permafrost_c = 0` currently. However, `test_hector.sh` confirms that the model runs with permafrost active.
* All parameters are exposed in the R interface.
* A copy of the 2001 paper permafrost code is archived in `misc/`.

## Still to do

Where will existence of permafrost be documented in v3?

### CODE REVIEW

**Comment by bpbond** [comment]
@kdorheim @dawnlwoodard  !!!!! Check it out, from 322007e6 :

permafrost

Progress 🕺

**Comment by dawnlwoodard** [comment]
@bpbond I remember there being a reason we separated static_c out into a pool originally and didn't just do a fraction but now I'm not remembering the justification for that. But I do know it's something we went back and forth on. Do you remember that rationale? and is it no longer relevant?

**Comment by bpbond** [comment]
🤷‍♂️ I don't remember. I couldn't see a good reason to keep it numerically, but at the same time that's part of why I am committing your code: so we have a close at hand, easy reference point in case we want to re-implement `static_c`.

File: test_hector.sh [hunk 1]
```diff
@@ -31,15 +31,21 @@ exit_loudly () {
    exit 1
 }
 
-
 # Run all INI files (the basic SSPs)
 echo "---------- Running: basic SSPs ----------"
 find ./inst/input -name "*.ini" -print0 | xargs -0 -n1 $HECTOR
 
+# Permafrost (off by default in v3) runs
+echo "---------- Running: permafrost ----------"
```
Comment (kdorheim) [inline test_hector.sh:39]: So this test checks to make sure can run without error?

**Review by kdorheim — CHANGES_REQUESTED** [review]
Some comments/questions to be addressed. I think talking though the data you've included with the PR would be helpful am also interested in Dawn's review

File: R/biome.R [hunk 2]
```diff
@@ -69,10 +74,12 @@ split_biome <- function(core,
         length(fveg_c) == length(new_biomes),
         length(fdetritus_c) == length(new_biomes),
         length(fsoil_c) == length(new_biomes),
+        length(fpermafrost_c) == length(new_biomes),
         length(fnpp_flux0) == length(new_biomes),
         sum(fveg_c) == 1, all(fveg_c > 0),
         sum(fdetritus_c) == 1, all(fdetritus_c > 0),
         sum(fsoil_c) == 1, all(fsoil_c > 0),
+        sum(fpermafrost_c) == 1, all(fpermafrost_c > 0),
```
Comment (dawnlwoodard) [inline R/biome.R:82]: this should be all(fpermafrost_c >= 0) since many biomes will not have permafrost in them

**Comment by dawnlwoodard** [comment]
Okay so when I go in and remake my pf paper figure (just with 2 SSPs for simplicity), I get some pretty different results for the lowest scenario (SSP126). Do we know why this is? And do we believe that it should happen so quickly? It's refreezing much faster than it did in my paper. I know refreezing was the trickiest dynamic we had to deal with in first designing it, so it's one I'm particularly looking at to make sure the behavior makes sense in this version. (Variable labels are not plotting for whatever reason - should be Permafrost C, Thawed Permafrost C, Atmospheric CH4, Atmospheric CO2, Global Mean Temperature)

image

Also, I don't see a way to get output for the different RH variables (rh_thawedp, rh_ch4, rh_soil, rh_det) that I used in my permafrost paper analysis. Probably not critical but wanted to flag this as I did find them useful outputs to be able to look at to double check behavior.

**Comment by dawnlwoodard** [comment]
Also, I get a pretty different ssp126 result if I run without biomes (the first figure was made by duplicating my paper methodology with a permafrost and non-permafrost biome). Which makes me wonder why the refreeze stops in the first, but not in this. That doesn't necessarily make sense to me.

image

File: src/ch4_component.cpp [hunk 3]
```diff
@@ -161,8 +161,13 @@ void CH4Component::run(const double runToDate) {
...
         << std::endl;
 
+    // Permafrost thaw produces CH4 emissions
+    #define PG_C_TO_TG_CH4 (1000.0 * 16.04 / 12.01)
+    const double rh_ch4 = core->sendMessage(M_GETDATA, D_RH_CH4).value(U_PGC_YR) * PG_C_TO_TG_CH4;
+
+    // Additional, background CH4 natural emissions
     const double ch4n = CH4N.value(U_TG_CH4);
-    const double emisTocon = (current_ch4em + ch4n) / UC_CH4.value(U_TG_PPBV);
+    const double emisTocon = (current_ch4em + rh_ch4 + ch4n) / UC_CH4.value(U_TG_PPBV);
```
[the remark this answers is not in the document]
Comment (dawnlwoodard) [reply src/ch4_component.cpp:170]: yes - Hector doesn't have any other methane respiration at the moment. But in the real world, of course, there is methane production from non-permafrost soils, so introducing this variable may open up questions about why we don't model other methane emissions yet.

File: misc/woodard_2021_permafrost/README.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+# woodard_2021_permafrost
+
```
[the remark this answers is not in the document]
Comment (dawnlwoodard) [reply misc/woodard_2021_permafrost/README.md:2]: yeah it's on Zenodo, so we can just reference that if that's preferable.

File: src/simpleNbox-runtime.cpp [hunk 5]
```diff
@@ -615,9 +773,18 @@ int SimpleNbox::calcderivs(double t, const double c[], double dcdt[]) const {
...
-
+  fluxpool rh_current = rh_fda_current + rh_fsa_current + rh_ftpa_co2_current;
+  fluxpool rh_ch4_current = rh_ftpa_ch4_current;
+  /*
+  cout << "permafrost_c = " << sum_map(permafrost_c) << endl;
+  cout << "thawed_permafrost = " << sum_map(thawed_permafrost_c) << endl;
+  cout << "thawed_permafrost = " << sum_map(thawed_permafrost_c) << endl;
+  cout << "rh_ftpa_co2_current = " << rh_ftpa_co2_current << endl;
+  cout << "rh_ftpa_ch4_current = " << rh_ftpa_ch4_current << endl;
+  */
```
Comment (dawnlwoodard) [inline src/simpleNbox-runtime.cpp:787]: can we remove these cout statements?

File: src/ch4_component.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bpbond) [reply src/ch4_component.cpp:170]: One answer would be it's respiration of brand new (that is, brand old long-sequestered) carbon in what are often very poorly drained environments. It's thus particularly important to model CH4.

File: src/simpleNbox-runtime.cpp [hunk 6]
```diff
@@ -373,10 +454,25 @@ void SimpleNbox::stashCValues(double t, const double c[]) {
...
+      fluxpool pf_refreeze_tp = yf * thawed_permafrost_c[biome].flux_from_fluxpool(fluxpool(y, U_PGC_YR));
+      fluxpool pf_refreeze_soil = yf * soil_c[biome].flux_from_fluxpool(fluxpool(z, U_PGC_YR));
+      // ...and update pools
+      permafrost_c[biome] = permafrost_c[biome] - pf_thaw + pf_refreeze_tp + pf_refreeze_soil;
+      thawed_permafrost_c[biome] = thawed_permafrost_c[biome] + pf_thaw - pf_refreeze_tp;
+      soil_c[biome] = soil_c[biome] - pf_refreeze_soil;
+    }
+    
     // Update litter from veg to soil and detritus
     fluxpool litter_flux = veg_c[biome] * (0.035 * yf);
```
[the remark this answers is not in the document]
Comment (bpbond) [reply src/simpleNbox-runtime.cpp:477]: It's the fraction of vegetation that transitions to litter each year. I don't know where it comes from.

(Note not a part of this PR.)

File: test_hector.sh (hunk 1 above)
Comment (bpbond) [reply test_hector.sh:39]: Right.

File: R/RcppExports.R [hunk 7]
```diff
@@ -1398,6 +1401,54 @@ SOIL_C <- function(biome = "") {
     .Call('_hector_SOIL_C', PACKAGE = 'hector', biome)
 }
 
+#' @describeIn carboncycle Permafrost C pool (`"Pg C"`)
+#' @param biome Name of biome (leave empty for global)
+#' @export
+PERMAFROST_C <- function(biome = "") {
```
[the remark this answers is not in the document]
Comment (bpbond) [reply R/RcppExports.R:1407]: Let's talk about this today?

File: R/biome.R [hunk 8]
```diff
@@ -50,6 +52,8 @@ create_biome <- function(core, biome,
 #'   biome. Defaults to the same value as \code{fveg_c}.
 #' @param fsoil_c Fractions of soil C distributed to each biome.
 #'   Defaults to the same value as \code{fveg_c}.
+#' @param fpermafrost_c Fraction of permafrost C distributed to each
```
[the remark this answers is not in the document]
Comment (bpbond) [reply R/biome.R:55]: Permafrost can occur in any biome—it's up to the user—so there's no "permafrost biome" per se.

File: R/biome.R (hunk 2 above)
Comment (bpbond) [reply R/biome.R:82]: Good catch—thanks.

**Comment by bpbond** [comment]
@kdorheim Draft permafrost vignette created in 01d9db83d7b0034163733336db9f3686752bf974  and is here: https://rpubs.com/bpbond/994470

**Comment by dawnlwoodard** [comment]
@kdorheim @bpbond the permafrost vignette doesn't mention running with biomes, which I believe is important to getting accurate permafrost results. Can this be added? I can share examples of how to do this if helpful.

**Pull request merged**